internal/utilities/siws/parser.go

```diff
@@ -0,0 +1,99 @@
+package siws
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+)
+
+// ParseSIWSMessage parses a raw SIWS message into an SIWSMessage struct,
+// performing robust checks to ensure correct formatting.
+func ParseSIWSMessage(raw string) (*SIWSMessage, error) {
+	lines := strings.Split(raw, "\n")
+
+	// Remove empty lines at the end or accidental trailing newlines.
+	// Some wallets or frameworks may add them.
+	var cleaned []string
+	for _, line := range lines {
+		l := strings.TrimSpace(line)
+		if l != "" {
+			cleaned = append(cleaned, l)
+		}
+	}
+	if len(cleaned) < 2 {
+		return nil, errors.New("message is too short or improperly formatted")
+	}
+
+	// 1) First line should match "<domain> wants you to sign in with your Solana account:"
+	// Use a regex to capture the domain.
+	domainRegex := regexp.MustCompile(`^([^ ]+)\s+wants you to sign in with your Solana account:$`)
```

Comment:
Please move all regexps to toplevel, so they're compiled once. Regex compilation is very expensive to be done on each call.